<a href="https://colab.research.google.com/github/jardinthee/codetree-TILs/blob/main/7_multilayer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 7장_멀티레이어 인공 신경망
히든레이어와 멀티레이어의 구조를 이해하고, 히든 레이어를 추가란 멀티레이어 인공 신경망 모델을 완성하기

In [1]:
#히든 레이어 실습

import tensorflow as tf
import pandas as pd

파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/boston.csv'
보스턴 = pd.read_csv(파일경로)

print(보스턴.columns)

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')


In [3]:
독립 = 보스턴[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax','ptratio', 'b', 'lstat']]
종속 = 보스턴[['medv']]

# 모델의 구조 만들기

In [4]:
X = tf.keras.layers.Input(shape=[13])
H = tf.keras.layers.Dense(10, activation = 'swish')(X)
Y = tf.keras.layers.Dense(1)(H)

model = tf.keras.models.Model(X,Y)
model.compile(loss = 'mse')

model.fit(독립,종속,epochs=100)

Epoch 1/100
16/16 [==============================] - 1s 2ms/step - loss: 7508.8574
Epoch 2/100
16/16 [==============================] - 0s 2ms/step - loss: 1844.9257
Epoch 3/100
16/16 [==============================] - 0s 2ms/step - loss: 675.9770
Epoch 4/100
16/16 [==============================] - 0s 2ms/step - loss: 481.8105
Epoch 5/100
16/16 [==============================] - 0s 3ms/step - loss: 339.5197
Epoch 6/100
16/16 [==============================] - 0s 2ms/step - loss: 205.5761
Epoch 7/100
16/16 [==============================] - 0s 2ms/step - loss: 130.5845
Epoch 8/100
16/16 [==============================] - 0s 2ms/step - loss: 97.5904
Epoch 9/100
16/16 [==============================] - 0s 2ms/step - loss: 85.3139
Epoch 10/100
16/16 [==============================] - 0s 2ms/step - loss: 75.8385
Epoch 11/100
16/16 [==============================] - 0s 2ms/step - loss: 68.9227
Epoch 12/100
16/16 [==============================] - 0s 2ms/step - loss: 62.9193
Epoch 13/100
16/

In [5]:
model.fit(독립,종속,epochs=1000,verbose=0)
model.fit(독립,종속, epochs=10)

Epoch 1/10
16/16 [==============================] - 0s 3ms/step - loss: 24.4281
Epoch 2/10
16/16 [==============================] - 0s 2ms/step - loss: 26.0881
Epoch 3/10
16/16 [==============================] - 0s 2ms/step - loss: 25.4057
Epoch 4/10
16/16 [==============================] - 0s 2ms/step - loss: 27.1420
Epoch 5/10
16/16 [==============================] - 0s 2ms/step - loss: 25.0988
Epoch 6/10
16/16 [==============================] - 0s 2ms/step - loss: 25.8018
Epoch 7/10
16/16 [==============================] - 0s 2ms/step - loss: 22.5152
Epoch 8/10
16/16 [==============================] - 0s 2ms/step - loss: 25.3234
Epoch 9/10
16/16 [==============================] - 0s 2ms/step - loss: 25.5505
Epoch 10/10
16/16 [==============================] - 0s 2ms/step - loss: 25.5198


In [6]:
model.fit(독립,종속,epochs=1000,verbose=0)
model.fit(독립,종속, epochs=10)

Epoch 1/10
16/16 [==============================] - 0s 2ms/step - loss: 27.0818
Epoch 2/10
16/16 [==============================] - 0s 2ms/step - loss: 24.0723
Epoch 3/10
16/16 [==============================] - 0s 2ms/step - loss: 21.9267
Epoch 4/10
16/16 [==============================] - 0s 2ms/step - loss: 27.0792
Epoch 5/10
16/16 [==============================] - 0s 2ms/step - loss: 22.5538
Epoch 6/10
16/16 [==============================] - 0s 2ms/step - loss: 25.6757
Epoch 7/10
16/16 [==============================] - 0s 2ms/step - loss: 24.0520
Epoch 8/10
16/16 [==============================] - 0s 3ms/step - loss: 24.0023
Epoch 9/10
16/16 [==============================] - 0s 2ms/step - loss: 23.2063
Epoch 10/10
16/16 [==============================] - 0s 2ms/step - loss: 24.8574


In [7]:
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 13)]              0         
                                                                 
 dense (Dense)               (None, 10)                140       
                                                                 
 dense_1 (Dense)             (None, 1)                 11        
                                                                 
Total params: 151 (604.00 Byte)
Trainable params: 151 (604.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


설명: 마지막 출력층(output_3)은 13개의 입력을 받아서 하나의 출력을 만드니까 10개의 항에 bias가 1개 추가되어서 10+1개의 가중치를 찾게 됨

두번째 층(dense_2)는 수식을 10개 만드는 것인데, 입력으로 받은 13개의 항에 마지막 bias 1개를 더해서 수식마다 14개의 가중치가 필요하므로 (13+1)*10 = 140개의 가중치를 찾는다고 해석됨

# 모델을 이용

In [8]:
print(model.predict(독립[:5]))
print(종속[:5])

1/1 [==============================] - 0s 110ms/step
[[30.414087]
 [23.03135 ]
 [29.161264]
 [34.26629 ]
 [32.9762  ]]
   medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2


# 붓꽃 품종 분류

이제 품종 분류도 멀티 레이어로 바꿔보겠습니다

과거의 데이터를 준비하기

In [13]:
파일경로 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
아이리스 = pd.read_csv(파일경로)

아이리스 = pd.get_dummies(아이리스)

독립 = 아이리스[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
종속 = 아이리스[['품종_setosa', '품종_versicolor','품종_virginica']]

print(독립.shape,종속.shape)

(150, 4) (150, 3)


모델의 구조 만들기

In [14]:
X = tf.keras.layers.Input(shape=[4])
H = tf.keras.layers.Dense(8,activation='swish')(X)
H = tf.keras.layers.Dense(8,activation='swish')(H)
H = tf.keras.layers.Dense(8,activation='swish')(H)
Y = tf.keras.layers.Dense(3,activation='softmax')(H)

model = tf.keras.models.Model(X,Y)
model.compile(loss='categorical_crossentropy',metrics='accuracy')

In [15]:
model.fit(독립,종속,epochs=10)

Epoch 1/10
5/5 [==============================] - 1s 3ms/step - loss: 2.2633 - accuracy: 0.3333
Epoch 2/10
5/5 [==============================] - 0s 4ms/step - loss: 1.9032 - accuracy: 0.3333
Epoch 3/10
5/5 [==============================] - 0s 4ms/step - loss: 1.6673 - accuracy: 0.3333
Epoch 4/10
5/5 [==============================] - 0s 3ms/step - loss: 1.4819 - accuracy: 0.3333
Epoch 5/10
5/5 [==============================] - 0s 4ms/step - loss: 1.3268 - accuracy: 0.4067
Epoch 6/10
5/5 [==============================] - 0s 4ms/step - loss: 1.1959 - accuracy: 0.5600
Epoch 7/10
5/5 [==============================] - 0s 3ms/step - loss: 1.0863 - accuracy: 0.6533
Epoch 8/10
5/5 [==============================] - 0s 3ms/step - loss: 0.9932 - accuracy: 0.6667
Epoch 9/10
5/5 [==============================] - 0s 4ms/step - loss: 0.9119 - accuracy: 0.6667
Epoch 10/10
5/5 [==============================] - 0s 3ms/step - loss: 0.8446 - accuracy: 0.6667


In [16]:
model.fit(독립,종속,epochs=100,verbose=0)
model.fit(독립,종속,epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0918 - accuracy: 0.9533
Epoch 2/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0893 - accuracy: 0.9667
Epoch 3/10
5/5 [==============================] - 0s 5ms/step - loss: 0.0862 - accuracy: 0.9600
Epoch 4/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0844 - accuracy: 0.9733
Epoch 5/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0807 - accuracy: 0.9800
Epoch 6/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0851 - accuracy: 0.9733
Epoch 7/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0792 - accuracy: 0.9733
Epoch 8/10
5/5 [==============================] - 0s 4ms/step - loss: 0.0918 - accuracy: 0.9733
Epoch 9/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0813 - accuracy: 0.9867
Epoch 10/10
5/5 [==============================] - 0s 3ms/step - loss: 0.0769 - accuracy: 0.9867


학습이 끝나면 학습이 잘됐는지 평가합니다

In [17]:
print(model.predict(독립[:5]))
print(종속[:5])

1/1 [==============================] - 0s 98ms/step
[[9.9905187e-01 9.4817230e-04 1.7856895e-08]
 [9.9816859e-01 1.8312938e-03 8.2370022e-08]
 [9.9833798e-01 1.6618944e-03 6.8862121e-08]
 [9.9658173e-01 3.4181981e-03 1.6644482e-07]
 [9.9900705e-01 9.9301001e-04 1.7640792e-08]]
   품종_setosa  품종_versicolor  품종_virginica
0          1              0             0
1          1              0             0
2          1              0             0
3          1              0             0
4          1              0             0
